internal/pkg/model/component.go

```diff
@@ -32,36 +32,23 @@ func NewComponentsProvider(index *storageapi.IndexComponents, logger log.Logger,
 	}
 }
 
-// RLock acquire read lock, before getting Components().
-// Update() is blocked until the read is finished.
-func (p *ComponentsProvider) RLock() {
-	p.updateLock.RLock()
-}
-
-// RUnlock release read lock.
-func (p *ComponentsProvider) RUnlock() {
-	p.updateLock.RUnlock()
-}
-
 func (p *ComponentsProvider) Components() *ComponentsMap {
+	p.updateLock.RLock()
+	defer p.updateLock.RUnlock()
 	return p.value
 }
 
 func (p *ComponentsProvider) Update(ctx context.Context) {
 	go func() {
 		startTime := time.Now()
 		p.logger.Infof("components update started")
-		p.updateLock.Lock()
-
-		defer p.updateLock.Unlock()
-		p.logger.Infof("components update: acquired lock")
-		lockTime := time.Now()
-
 		ctx, cancel := context.WithTimeout(ctx, ComponentsUpdateTimeout)
 		defer cancel()
 		if index, err := p.index(ctx); err == nil {
+			p.updateLock.Lock()
+			defer p.updateLock.Unlock()
 			p.value = NewComponentsMap(index.Components)
-			p.logger.Infof("components update finished | %s / %s", time.Since(startTime), time.Since(lockTime))
+			p.logger.Infof("components update finished | %s", time.Since(startTime))
```

Comment:
Pri aktualizacii komponent sa nastavil zamok, vsetky requests museli pockat kym sa aktualizacia dokonci.
A naopak, update komponent musel pockat, kym dobehnu requests, ktore prisli skor.
Bolo to preto, aby bolo garantovane, ze pocas celeho requestu sa `components` nezmenia.

Teraz sa zamok pouziva iba na nevyhnutny cas, na vymenu hodnot (nie na cele nacitanie komponent).

------

Je to mozne preto, pretoze pre kazdy request uz vytvarame vlastny dependencies container `api.server.templates.dependencies.ForPublicRequest` (`ForProjectRequest`).

A tam si to vieme pri prvom pouziti pekne cachovat (uz to tam je):
https://github.com/keboola/keboola-as-code/blob/4c23574e44ee1e27a49e3dbcc7ce1091a9eaf07a/internal/pkg/api/server/templates/dependencies/dependencies.go#L246-L251

Tj. pri prvom zavolani sa `components` nacitaju z `dependencies.ForServer`, tj. z `ComponentProvider.Components()` a tato hodnota sa pouziva pocas celeho trvania requestu.

Pocas requestu moze nastat aktualizacia v `ComponentsProvider`, no request to neovplyvnuje.
Ked dobehne request a nastal `update` components, tak na staru hodnotu neodkazuje uz ziaden kod, a `garbage collector` ju z pamate odstrani.